SEP-24: Communicate KYC status (#739)
stellar/stellar-protocol: ecosystem/sep-0024.md

Name | Type | Description
-----|------|------------
`callback` | string | (optional) A URL that the anchor should `POST` a JSON message to when the user successfully completes the interactive flow. Can also be set to `postMessage`. 
`on_change_callback` | string | (optional) A URL that the anchor should `POST` a JSON message to when the `status` or `kyc_verified` properties change. Can also be set to `postMessage`.

**`callback` details**

If the wallet wants to be notified that the user has completed the anchor's interactive flow (either success or failure), it can add this parameter to the URL. If the user abandons the process, the anchor does not need to report anything to the wallet. If the `callback` value is a URL, the anchor must `POST` to it with a JSON message as the body. If `callback=postMessage` is passed, the anchor must post a JSON message to `window.opener` via the Javascript [`Window.postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) method. If `window.opener` is undefined, the message must be posted to `window.parent` instead.

**Differences between `callback` and `on_change_callback`**

> Anchors should only send responses to `callback` when the user is finished with the interactive flow, and it's safe for the wallet to close the window. It should only be called zero or one time.

Reviewer: ```suggestion
Anchors may make _at most one request_ containing a `/transaction` response body to `callback` when the user is finished with the interactive flow _and_ the anchor has verified that it is safe for the wallet to close the popup window. 
```

Name | Type | Description
-----|------|------------

> `callback` | string | (optional) A URL that the anchor should `POST` a JSON message to when the `status` property changes, or when the `kyc_verified` property changes and the `status` property is incomplete or pending. Can also be set to `postMessage`.

Reviewer: ```suggestion
`callback` | string | (optional) A URL that the anchor should `POST` a JSON message to when the `status` or `kyc_verified` properties change. Can also be set to `postMessage`.
```

Reviewer 2: Prior to this change if the client uses `postMessage` as the callback they expect to get a single call to postMessage when the interactive flow has changed. With this change an implementer might think they should call postMessage during the interactive flow resulting in multiple `postMessage` calls. It's likely because of the previous definition that a client has implemented their application to close the window containing the interactive flow when the first postMessage is received.

I think because of this we need to retain some of the original language so that when postMessage is used the first call to postMessage indicates not only a change in status or properties but that the interactive flow has completed.

Reviewer: +1 to Leigh's comment, I didn't look at the original language before making my comment.

Author: @leighmcculloch I'm not quite understanding what you're suggesting? It seems like our proposed change will be a breaking one for your specific case, and I don't think there's any way around it. Unless you're saying that postMessage would behave differently than callback URLs?

Reviewer: @morleyzhi I think what Leigh is saying is that the original purpose to the `postMessage` call was for the client to close the window. There's no reason to remove this language, since the change we're making is additive. The first `postMessage` callback can signal that the flow is complete AND subsequent callbacks can signal a change in `status` or `kyc_verified`.

Reviewer: Note that when the flow completes, the `status` will likely change too, so the distinction between the two purposes is actually not needed from a functional standpoint.

Author: The problem is that there may be some status changes before the flow is complete: the incomplete -> pending_kyc one comes to mind.

Author: If we're concerned about backwards compatibility, we can add a second callback property.

Reviewer: That may be necessary, because if you look at the SEP, `callback` is stated to be used to signal the end of the interactive flow all over the place. Going back on this would cause issues for wallets expecting a single callback request.

Reviewer 2: `callback_on_change` sounds good to me.

> 

Reviewer 2: Immediately above here is a good section that details the callback parameter and what data is posted, etc. I think we need a similar section for the new parameter. You could mostly copy paste it and note how the `on_change_callback` is posted to on every change.

As an alternative to using the `callback` parameter, the wallet can poll the transaction endpoint [`/transaction`](#single-historical-transaction) with the request `id` to check the status of the request.

**Differences between `callback` and `on_change_callback`**

Anchors should only send responses to `callback` when the user is finished with the interactive flow, and it's safe for the wallet to close the window. It should only be called zero or one time.

`on_change_callback` should be called each time the `status` or `kyc_verified` properties of the transaction change. It can be called any number of times.

> 

Reviewer 2: This document appears to use single line separation between sections. Let's drop the double.
```suggestion
```

> `on_change_callback` should be called each time the `status` or `kyc_verified` properties of the transaction should change. It can be called any number of times.

Reviewer 2: ```suggestion
`on_change_callback` should be called each time the `status` or `kyc_verified` properties of the transaction change. It can be called any number of times.
```

`kind` | string | `deposit` or `withdrawal`.
`status` | string | Processing status of deposit/withdrawal.
`status_eta` | number | (optional) Estimated number of seconds until a status change is expected.

> `kyc_verified` | boolean | (optional) Whether or not the user's KYC information has been verified by the anchor.

Reviewer 3: Should we add a qualifications here that KYC has been verified for this specific transaction? The current language might conflate the status of the user with the status of the transaction. KYC requirements can change and different tiers can be invoked in the future so a wallet should make a concise decision wrt what this status means in the context of a user.

Author: Yeah, that's a good idea.

> `kyc_verified` | boolean | (optional) Whether or not the deposit/withdraw subject has been verified by the anchor.

Reviewer: ```suggestion
`kyc_verified` | boolean | (optional) Whether or not the user's KYC information has been verified by the anchor.
```

`kind` | string | `deposit` or `withdrawal`.
`status` | string | Processing status of deposit/withdrawal.
`status_eta` | number | (optional) Estimated number of seconds until a status change is expected.

> `kyc_verified` | boolean | (optional) Whether or not the anchor has verified the user's KYC information and has authorized them for this transaction.

Reviewer 2: We should avoid the term authorized here, because the term has particular meaning in payments products and our overloaded use of it may cause confusion for folks in financial industries. I think its use here is unnecessary anyway and can be simply deleted.
```suggestion
`kyc_verified` | boolean | (optional) True if the anchor has verified the user's KYC information for this transaction.
```

`status` should be one of:

* `incomplete` -- there is not yet enough information for this transaction to be initiated. Perhaps the user has not yet entered necessary info in an interactive flow.
* `pending_kyc` -- the user has submitted their KYC information and the anchor is verifying it.
* `pending_user_transfer_start` -- the user has not yet initiated their transfer to the anchor. This is the necessary first step in any deposit or withdrawal flow.
* `pending_external` -- deposit/withdrawal has been submitted to external network, but is not yet confirmed. This is the status when waiting on Bitcoin or other external crypto network to complete a transaction, or when waiting on a bank transfer.

> * `pending_anchor` -- deposit/withdrawal is being processed internally by anchor. This can also be used when the anchor must verify KYC information prior to deposit/withdrawal.

Reviewer 3: Are we keeping the KYC language on `pending_anhcor`?

Reviewer 2: It does sound like pending_anchor already does the thing we want.

Reviewer: Good catch Tomer. This second sentence should be removed, however existing anchors will likely continue to use `pending_anchor` while verifying KYC. No change for them is needed unless they work with a client that wants to see the `pending_kyc` status.

Reviewer: @leighmcculloch it doesn't because its a wider umbrella then just KYC. SEP-24 needs have a way to communicate that the anchor is specifically working on KYC validation

Reviewer 3: @JakeUrban, what other states are under that wider umbrela of `pending_anchor`?

Reviewer: `pending_anchor` is not completely covered by sub-statuses, really there is no such thing other than this one case.

Meaning, `pending_anchor` historically has been a catch-all for "things the anchor needs to do that it doesn't have a better status for". Now that there's `pending_kyc`, the anchor should use that whenever its relevant in place of `pending_anchor`, however `pending_anchor` will still be necessary for other situations.

Reviewer: There are no other statuses that could be seen as substitutes for. `pending_anchor`. Its possible some anchors use `pending_anchor` when really its `pending_external`, but thats a minor thing and has no affect on functionality.

Reviewer 3: @JakeUrban can you give some examples of these other situations that `pending_anchor` will be necessary for?

It is my understanding that this PR attempts to de-complect kyc verification from other potential scenarios in which the wallet is waiting on the anchor. I just don't have a full understanding of what those scenarios are. 

My mental model of `pending_anchor` has been that someone needs to take a look at the KYC information, transaction details and make a yes/no decision. If you take that way from `pending_anchor` - what is left? 

This is a breaking change. We need to have strong conviction here.

Author: We'll use `pending_anchor` for now.

Reviewer: Other than KYC, `pending_anchor` is used when 

- Polaris is constructing the Stellar transaction envelope to be submitted to the network
- Polaris has detected a withdrawal payment to its distribution account and is sending the funds off chain

Directly after the first case listed, the status will be `pending_stellar`. Directly after the second case listed, the status will be `pending_external` if still pending or `completed` if the user received the funds off-chain.

Reviewer 3: > Polaris has detected a deposit payment to the anchor's bank account and is constructing the Stellar transaction envelope to be submitted to the network

Reviewer 3: Jake, this is super helpful. In both these specific cases it sounds like `pending_anchor` is a very short ephemeral status - can we just skip it?

Though I imagine that with multi-sig distribution transactions your first example can take a longer...

Reviewer: > can we just skip it?

From a UX perspective, the same rationale could be used for `pending_stellar`, since it's also very temporary. The answer depends on if you want clients or users to know either of those two things happen. If not, you can just notify of success or failure when it happens. Until then, it would be in `pending_user_transfer_start`.

If the transaction requires signatures from distributed systems, then being `pending_anchor` is much more significant since it could take a long time.

Author: To get this moving, I removed pending_kyc. Anchors that don't support any change can use `pending_anchor` for now.
